lib/virtual-fs/src/union_fs.rs

```diff
@@ -158,17 +158,19 @@ impl UnionFileSystem {
     }
 
     fn read_dir_internal(&self, path: &Path) -> Result<ReadDir> {
-        let path = path.to_string_lossy();
+        let path_str = path.to_string_lossy();
 
         let mut ret = None;
-        for (path, mount) in filter_mounts(&self.mounts, path.as_ref()) {
-            match mount.fs.read_dir(Path::new(path.as_str())) {
+        for (p, mount) in filter_mounts(&self.mounts, path_str.as_ref()) {
+            match mount.fs.read_dir(Path::new(p.as_str())) {
                 Ok(dir) => {
                     if ret.is_none() {
                         ret = Some(Vec::new());
                     }
                     let ret = ret.as_mut().unwrap();
-                    for sub in dir.flatten() {
+                    for mut sub in dir.flatten() {
+                        let sub_path: PathBuf = sub.path.components().skip(1).collect();
+                        sub.path = PathBuf::from(&mount.path).join(&sub_path);
```

Comment:
Does this code need to handle empty sub-paths?

Reply:
`sub` contains entries with this format: `/path/to/entry`. The smallest number of components `sub` can have is when you read the root directory and get something like this: `/foo`. so `sub_path` will contain `foo`. The `read_dir` must return `/` for `sub_path` to become empty which isn't a valid value.